paddle/fluid/prim/api/auto_code_generated/tensor_operants_gen.py

```diff
@@ -0,0 +1,273 @@
+
+"""
+
+
+eager_header_end = """};
+
+}  // namespace prim
+}  // namespace paddle
+
+"""
+
+
+eager_source_include = """// Generated by paddle/fluid/prim/api/auto_code_generated/tensor_operants_gen.py
+
+#include "paddle/fluid/prim/utils/eager/eager_tensor_operants.h"
+
+#include "paddle/fluid/eager/api/generated/eager_generated/forwards/dygraph_functions.h"
+
+"""
+
+
+eager_source_start = """
+namespace paddle {
+
+namespace prim {
+
+"""
+
+
+eager_source_end = """
+}  // namespace prim
+}  // namespace paddle
+
+"""
+
+
+static_header_include = """// Generated by paddle/fluid/prim/api/auto_code_generated/tensor_operants_gen.py
+
+#pragma once
+
+#include "paddle/phi/api/include/operants_base.h"
+#include "paddle/phi/api/include/tensor.h"
+#include "paddle/phi/core/macros.h"
+
+"""
+
+static_header_start = """
+namespace paddle {
+
+namespace prim {
+
+using Tensor = paddle::experimental::Tensor;
+using TensorOperantsBase = paddle::operants::TensorOperantsBase;
+
+class StaticTensorOperants : public TensorOperantsBase {
+ private:
+  DISABLE_COPY_AND_ASSIGN(StaticTensorOperants);
+
+ public:
+  StaticTensorOperants() = default;
+
+"""
+
+
+static_header_end = """};
+
+}  // namespace prim
+}  // namespace paddle
+
+"""
+
+
+static_source_include = """// Generated by paddle/fluid/prim/api/auto_code_generated/tensor_operants_gen.py
+
+#include "paddle/fluid/prim/utils/static/static_tensor_operants.h"
+
+#include "paddle/fluid/prim/api/generated_prim/prim_generated_api.h"
+#include "paddle/fluid/prim/utils/static/desc_tensor.h"
+
+"""
+
+
+static_source_start = """
+namespace paddle {
+
+namespace prim {
+using DescTensor = paddle::prim::DescTensor;
+
+"""
+
+
+static_source_end = """
+}  // namespace prim
+}  // namespace paddle
+
+"""
+
+
+def generate_tensor_operants_api(
+    api_yaml_path,
+    eager_header_path,
+    eager_source_path,
+    static_header_path,
+    static_source_path,
+    api_prim_path,
+):
+    apis = []
+
+    for each_api_yaml in api_yaml_path:
+        with open(each_api_yaml, 'r') as f:
+            api_list = yaml.load(f, Loader=yaml.FullLoader)
+            if api_list:
+                apis.extend(api_list)
+
+    eager_header_file = open(eager_header_path, 'w')
+    eager_source_file = open(eager_source_path, 'w')
+    static_header_file = open(static_header_path, 'w')
+    static_source_file = open(static_source_path, 'w')
+
+    eager_header_file.write(eager_header_include)
+    eager_header_file.write(eager_header_start)
+    eager_source_file.write(eager_source_include)
+    eager_source_file.write(eager_source_start)
+    static_header_file.write(static_header_include)
+    static_header_file.write(static_header_start)
+    static_source_file.write(static_source_include)
+    static_source_file.write(static_source_start)
+
+    with open(api_prim_path, 'rt') as f:
+        api_prims = yaml.safe_load(f)
+        # white list temporarily
+        api_prims = ('add', 'subtract', 'multiply', 'divide')
```

Comment:
第一句的api_prims赋值有什么作用，另外`api_prims = ('add', 'subtract', 'multiply', 'divide')`这个应该会多次做出修改，是不是可以作为一个全局变量放在外边，方便修改维护

Reply:
现在只重载了四则运算，`api_prims = ('add', 'subtract', 'multiply', 'divide')` 只是一个短期的白名单。后续会直接将这行代码删掉，根据 `api_prims = yaml.safe_load(f)` 生成 prim 的 api。

`api_prims = ('add', 'subtract', 'multiply', 'divide')` is a temporary whitelist solution because tensor operants only overload arithmetic operators. In the future, this line of code will be deleted and generate prim api according to `api_prims = yaml.safe_load(f)`